### PR TITLE
C++ port writer

### LINKED ISSUE #191: Implement C++ ports
Implement C++ code gen for ports.

### PR DESCRIPTION
**Pull request by tiffany1618**
Closes #191, #192

FPP changes:
- Add `PortCppWriter` for writing C++ for port definitions
- Add unit tests for C++ port writer
- Update unit tests for `fpp-depend` and `fpp-filenames`
- Add support for writing un-nested string classes in `StringCppWriter` 

Changes from Python port autocoder:
- Port buffer class only generated for ports without return types, since it is only used for port serialization
- Port string classes are wrapped the namespace `[port type name]PortStrings` to avoid polluting the global namespace

Temporary fixes for compatibility with Python component autocoder:
- Typedefs are generated to support string type names derived from the string parameter names, e.g. `typedef StringSize[size] [param name]String;`

### CODE REVIEW

**Comment by bocchino** [comment]
Looks good! For this:

>Temporary fixes for compatibility with Python component autocoder:

>Generated string class names are derived from the name of the corresponding string parameter instead of the string size (<paramName>String instead of StringSize<size>)
Typedefs are generated for strings of the same size to avoid duplicating string classes

I had envisioned that we would keep generating string class names according to the string size (since that's what we eventually want to do) and add typedefs with the names of the string parameters to keep the component autocoder happy. Then later we can delete the typedefs. What do you think?

**Comment by bocchino** [comment]
A related point is that I see changes to the struct and array code gen. I don't think this is necessary. I think we can just add the necessary typedefs to the port code gen.

**Comment by tiffany1618** [comment]
> I had envisioned that we would keep generating string class names according to the string size (since that's what we eventually want to do) and add typedefs with the names of the string parameters to keep the component autocoder happy. Then later we can delete the typedefs. What do you think?

I agree, this makes more sense.

> A related point is that I see changes to the struct and array code gen. I don't think this is necessary. I think we can just add the necessary typedefs to the port code gen.

This was actually unrelated to the string class renaming. Arrays and structs both generate nested string classes, but ports do not, so the string class writer previously only generated nested classes. When I added generation for the un-nested string classes, I renamed the `write` function

**Comment by bocchino** [comment]
> Arrays and structs both generate nested string classes, but ports do not

OK, thanks. I guess the string classes aren't nested in the port classes because there are two port classes. However, we need some kind of name qualification for the string classes. We can't put generic names like 'StringSize80' in the global namespace if we are generating one of these classes for each port type. Can we put the definitions in a namespace like `[port type name]Strings`?

**Comment by tiffany1618** [comment]
>Can we put the definitions in a namespace like `[port type name]Strings`?

Yes, that sounds good.

**Comment by bocchino** [comment]
OK, great! Maybe the namespace should be `[port type name]PortStrings`. This is consistent with the other code gen that adds `Port` to the port type name in the generated symbols.

**Comment by tiffany1618** [comment]
Ok, done! The string typedefs are still in the global namespace, since the Python port autocoder was generating its string classes there.

File: compiler/lib/src/main/scala/codegen/CppWriter/PortCppWriter.scala [hunk 1]
```diff
@@ -0,0 +1,624 @@
...
+  }
+
+  private def getHppIncludes: CppDoc.Member = {
+    val serializableHeader = data.returnType match {
+      case Some(_) => Nil
+      case None => List("Fw/Types/Serializable.hpp")
+    }
+    val standardHeaders = (
+      List(
+        "cstdio",
```
Comment (bocchino) [inline compiler/lib/src/main/scala/codegen/CppWriter/PortCppWriter.scala:166]: cstdio and cstring should be system headers <...> not standard headers "...".

File: compiler/tools/fpp-to-cpp/test/port/ReturnTypePortAc.ref.cpp [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+    return this->m_func(this->m_comp, this->m_portNum, u);
+  }
+
+#if FW_PORT_SERIALIZATION == 1
+
+  Fw::SerializeStatus InputReturnTypePort ::
+    invokeSerial(Fw::SerializeBufferBase& _buffer)
+  {
+    // For ports with a return type, invokeSerial is not used
```
Comment (bocchino) [inline compiler/tools/fpp-to-cpp/test/port/ReturnTypePortAc.ref.cpp:63]: clang is giving me an unused parameter warning when compiling this code. F Prime currently turns off these warnings, but I think we should make our generated code comply with them going forward. Let's add the following line to the function body to suppress the warning:
```
(void) _buffer;
```

**Review by bocchino — COMMENTED** [review]
Looks great! Just a couple of comments.

Also, I think this PR does not close #193. So let's remove "closes #193" from the initial comment.

File: compiler/lib/src/main/scala/codegen/CppWriter/PortCppWriter.scala (hunk 1 above)
Comment (bocchino) [reply compiler/lib/src/main/scala/codegen/CppWriter/PortCppWriter.scala:166]: Compare how we handle system headers in the other Cpp writers.

Comment (tiffany1618) [reply compiler/lib/src/main/scala/codegen/CppWriter/PortCppWriter.scala]: Fixed!

File: compiler/tools/fpp-to-cpp/test/port/ReturnTypePortAc.ref.cpp (hunk 2 above)
Comment (tiffany1618) [reply compiler/tools/fpp-to-cpp/test/port/ReturnTypePortAc.ref.cpp:63]: Fixed!

**Review by bocchino — APPROVED** [review]
Looks good!

**Pull request merged**